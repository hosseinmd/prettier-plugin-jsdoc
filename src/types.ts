import { ParserOptions } from "prettier";

export interface JsdocOptions {
  jsdocSpaces: number;
  jsdocPrintWidth?: number;
  jsdocDescriptionWithDot: boolean;
  jsdocDescriptionTag: boolean;
  jsdocVerticalAlignment: boolean;
  jsdocKeepUnParseAbleExampleIndent: boolean;
  /** default is true */
  jsdocSingleLineComment: boolean;
  jsdocSeparateReturnsFromParam: boolean;
  jsdocSeparateTagGroups: boolean;
  jsdocAddDefaultToDescription: boolean;
  jsdocCapitalizeDescription: boolean;
  tsdoc: boolean;
}

export interface AllOptions extends ParserOptions, JsdocOptions {}

type LocationDetails = { line: number; column: number };
type Location = { start: LocationDetails; end: LocationDetails };

export type PrettierComment = {
  type: "CommentBlock" | "Block";
  value: string;
  start: number;
  end: number;
  loc: Location;
};

export type Token = {
  type:
    | "CommentBlock"
    | "Block"
    | {
        label: string; // "function" | "name";
        keyword?: string;
        beforeExpr: boolean;
        startsExpr: boolean;
        rightAssociative: boolean;
        isLoop: boolean;
        isAssign: boolean;
        prefix: boolean;
        postfix: boolean;
        binop: null;
      };
  value: string;
  start: number;
  end: number;
  loc: Location;
};

export type AST = {
  start: number;
  end: number;
  loc: Location;
  errors: [];
  program: {
    type: "Program";
    start: number;
    end: number;
    loc: [];
    sourceType: "module";
    interpreter: null;
    body: [];
    directives: [];
  };
  comments: PrettierComment[];
  tokens: Token[];
};
