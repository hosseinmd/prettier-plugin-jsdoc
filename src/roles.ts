import {
  ABSTRACT,
  ASYNC,
  AUGMENTS,
  AUTHOR,
  CALLBACK,
  CATEGORY,
  CLASS,
  CONSTANT,
  DEFAULT,
  DEPRECATED,
  DESCRIPTION,
  EXAMPLE,
  EXTENDS,
  EXTERNAL,
  FILE,
  FIRES,
  FLOW,
  FUNCTION,
  MEMBER,
  MEMBEROF,
  OVERRIDE,
  PARAM,
  PRIVATE,
  PROPERTY,
  PROVIDES_MODULE,
  RETURNS,
  SEE,
  SINCE,
  TEMPLATE,
  THROWS,
  TODO,
  TYPE,
  TYPEDEF,
  VERSION,
  YIELDS,
} from "./tags";

const TAGS_SYNONYMS = {
  // One TAG TYPE can have different titles called SYNONYMS.  We want
  // to avoid different titles in the same tag so here is map with
  // synonyms as keys and tag type as value that we want to have in
  // final jsDoc.
  arg: PARAM,
  argument: PARAM,
  const: CONSTANT,
  constructor: CLASS,
  defaultvalue: DEFAULT,
  desc: DESCRIPTION,
  emits: FIRES,
  examples: EXAMPLE,
  exception: THROWS,
  fileoverview: FILE,
  func: FUNCTION,
  host: EXTERNAL,
  method: FUNCTION,
  overview: FILE,
  params: PARAM,
  prop: PROPERTY,
  return: RETURNS,
  var: MEMBER,
  virtual: ABSTRACT,
  yield: YIELDS,
};

const TAGS_NAMELESS = [
  CATEGORY,
  DESCRIPTION,
  EXAMPLE,
  EXTENDS,
  RETURNS,
  THROWS,
  TODO,
  YIELDS,
  DEPRECATED,
  OVERRIDE,
  SINCE,
];

const TAGS_TYPELESS = [DESCRIPTION, EXAMPLE, TODO, DEPRECATED, OVERRIDE, SINCE];

const TAGS_DESCRIPTION_NEEDED = [CATEGORY, DESCRIPTION, EXAMPLE, SINCE, TODO];
const TAGS_NEED_FORMAT_DESCRIPTION = [
  CATEGORY,
  DESCRIPTION,
  PARAM,
  PROPERTY,
  RETURNS,
  THROWS,
  TODO,
  TYPE,
  TYPEDEF,
  YIELDS,
  DEPRECATED,
  OVERRIDE,
];
const TAGS_TYPE_NEEDED = [
  EXTENDS,
  PARAM,
  PROPERTY,
  RETURNS,
  THROWS,
  TYPE,
  TYPEDEF,
  YIELDS,
];

const TAGS_VERTICALLY_ALIGN_ABLE = [
  EXTENDS,
  PARAM,
  PROPERTY,
  RETURNS,
  THROWS,
  TYPE,
  TYPEDEF,
  YIELDS,
];

const TAGS_IS_CAMEL_CASE = [PROVIDES_MODULE];
const TAGS_GROUP = [CALLBACK, TYPEDEF];

const TAGS_ORDER = [
  PROVIDES_MODULE,
  FLOW,
  ASYNC,
  PRIVATE,
  MEMBEROF,
  VERSION,
  AUTHOR,
  DEPRECATED,
  SINCE,
  CATEGORY,
  DESCRIPTION,
  EXAMPLE,
  ABSTRACT,
  AUGMENTS,
  CONSTANT,
  DEFAULT,
  EXTERNAL,
  FILE,
  FIRES,
  TEMPLATE,
  FUNCTION,
  CLASS,
  EXTENDS,
  MEMBER,
  TYPEDEF,
  TYPE,
  PROPERTY,
  CALLBACK,
  PARAM,
  YIELDS,
  RETURNS,
  THROWS,
  "other",
  SEE,
  TODO,
];

export {
  TAGS_DESCRIPTION_NEEDED,
  TAGS_IS_CAMEL_CASE,
  TAGS_NAMELESS,
  TAGS_NEED_FORMAT_DESCRIPTION,
  TAGS_GROUP,
  TAGS_ORDER,
  TAGS_SYNONYMS,
  TAGS_TYPE_NEEDED,
  TAGS_TYPELESS,
  TAGS_VERTICALLY_ALIGN_ABLE,
};
