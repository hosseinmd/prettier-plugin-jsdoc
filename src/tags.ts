const ABSTRACT = "abstract";
const ASYNC = "async";
const AUGMENTS = "augments";
const AUTHOR = "author";
const CALLBACK = "callback";
const CATEGORY = "category";
const CLASS = "class";
const CONSTANT = "constant";
const DEFAULT = "default";
const DEPRECATED = "deprecated";
const DESCRIPTION = "description";
const EXAMPLE = "example";
const EXTENDS = "extends";
const EXTERNAL = "external";
const FILE = "file";
const FIRES = "fires";
const FLOW = "flow";
const FUNCTION = "function";
const MEMBER = "member";
const MEMBEROF = "memberof";
const OVERRIDE = "override";
const PARAM = "param";
const PRIVATE = "private";
const PROPERTY = "property";
const PROVIDES_MODULE = "providesModule";
const RETURNS = "returns";
const SEE = "see";
const SINCE = "since";
const TEMPLATE = "template";
const THROWS = "throws";
const TODO = "todo";
const TYPE = "type";
const TYPEDEF = "typedef";
const VERSION = "version";
const YIELDS = "yields";

export {
  ABSTRACT,
  ASYNC,
  AUGMENTS,
  AUTHOR,
  CALLBACK,
  CATEGORY,
  CLASS,
  CONSTANT,
  DEFAULT,
  DEPRECATED,
  DESCRIPTION,
  EXAMPLE,
  EXTENDS,
  EXTERNAL,
  FILE,
  FIRES,
  FLOW,
  FUNCTION,
  MEMBER,
  MEMBEROF,
  OVERRIDE,
  PARAM,
  PRIVATE,
  PROPERTY,
  PROVIDES_MODULE,
  RETURNS,
  SEE,
  SINCE,
  TEMPLATE,
  THROWS,
  TODO,
  TYPE,
  TYPEDEF,
  VERSION,
  YIELDS,
};
