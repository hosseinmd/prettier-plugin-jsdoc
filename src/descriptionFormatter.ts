import { DESCRIPTION } from "./tags";
import { Comment } from "comment-parser";
import { JsdocOptions } from "./types";
import {
  capitalizer,
  trimEmptyLines,
  trimTrailingSpaces,
  tryFormat,
} from "./utils";

interface FormatOptions {
  firstLinePrintWidth?: number;
}

/**
 * Trim, make single line with capitalized text. Insert dot if flag for it is
 * set to true and last character is a word character
 *
 * @private
 */
function formatDescription(
  text: string,
  options: JsdocOptions,
  formatOptions?: FormatOptions,
): string {
  text = trimTrailingSpaces(text);
  text = trimEmptyLines(text);
  text = capitalizer(text);

  if (!text) return text;

  const { printWidth } = options;
  let { firstLinePrintWidth = printWidth } = formatOptions ?? {};
  if (firstLinePrintWidth > printWidth) {
    firstLinePrintWidth = printWidth;
  }

  const shortFirstLine =
    firstLinePrintWidth > 1 &&
    firstLinePrintWidth < printWidth &&
    !needsItsOwnLine(text);

  if (shortFirstLine) {
    const prefix = "Z".repeat(printWidth - firstLinePrintWidth - 1) + " ";
    text = prefix + text;

    text =
      tryFormat(text, {
        ...options,
        parser: "markdown",
        proseWrap: "always",
      }) ?? text;

    text = text.replace(/^Z+[ ]*/, "");
  } else {
    text =
      tryFormat(text, {
        ...options,
        parser: "markdown",
        proseWrap: "always",
      }) ?? text;
  }

  return trimEmptyLines(text);
}

function needsItsOwnLine(markdown: string): boolean {
  return (
    // list
    /^\s*(?:\d+[.)]|[*])[\s-]/.test(markdown) ||
    // code block
    /^\s*```.*[\r\n]/.test(markdown) ||
    // heading
    /^\s*(?:#{1,6}.*|[^\s=-].*\n(?:===+|---+))/.test(markdown)
  );
}

function convertCommentDescToDescTag(parsed: Comment): void {
  if (!parsed.description) {
    return;
  }

  const Tag = parsed.tags.find(({ tag }) => tag.toLowerCase() === DESCRIPTION);
  let { description = "" } = Tag || {};

  description += parsed.description;

  if (Tag) {
    Tag.description = description;
  } else {
    parsed.tags.push({ tag: DESCRIPTION, description } as any);
  }
}

export { convertCommentDescToDescTag, formatDescription };
