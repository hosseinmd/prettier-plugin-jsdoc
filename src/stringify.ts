import { Tag } from "comment-parser";
import { formatDescription } from "./descriptionFormatter";
import { DESCRIPTION, EXAMPLE, TODO } from "./tags";
import {
  TAGS_NEED_FORMAT_DESCRIPTION,
  TAGS_VERTICALLY_ALIGN_ABLE,
} from "./roles";
import { JsdocOptions } from "./types";
import {
  trimEmptyLines,
  trimIndentation,
  trimTrailingSpaces,
  prefixLinesWith,
  getFirstLine,
  getLastLine,
  tryFormat,
} from "./utils";

function stringifyJSDoc(tags: Tag[], options: JsdocOptions): string {
  const {
    printWidth,
    jsdocSpaces,
    jsdocVerticalAlignment,
    jsdocDescriptionTag,
    jsdocKeepUnParseAbleExampleIndent,
  } = options;

  const shouldAlign = (tag: string) =>
    jsdocVerticalAlignment && TAGS_VERTICALLY_ALIGN_ABLE.includes(tag);

  const toAlignTags = tags.filter(({ tag }) => shouldAlign(tag));
  const MAX_TAG_LENGTH = Math.max(
    ...toAlignTags.map(({ tag }) => (tag || "").length),
  );
  const MAX_TYPE_LENGTH = Math.max(
    ...toAlignTags.map(({ type }) => (type || "").length),
  );
  const MAX_NAME_LENGTH = Math.max(
    ...toAlignTags.map(({ name }) => (name || "").length),
  );

  const TYPE_OFFSET = "@".length + MAX_TAG_LENGTH + jsdocSpaces;
  const NAME_OFFSET =
    TYPE_OFFSET +
    (MAX_TYPE_LENGTH ? MAX_TYPE_LENGTH + "{}".length + jsdocSpaces : 0);
  const DESC_OFFSET =
    NAME_OFFSET + (MAX_NAME_LENGTH ? MAX_NAME_LENGTH + jsdocSpaces : 0);

  interface TagStringifyResult {
    value: string;
    insertLineBefore?: boolean;
    insertLineAfter?: boolean;
  }
  function stringifyTag({
    tag,
    type,
    name,
    description,
  }: Tag): TagStringifyResult {
    if (tag === DESCRIPTION && !jsdocDescriptionTag) {
      return {
        value: formatDescription(description, options),
        insertLineAfter: true,
      };
    } else if (tag === EXAMPLE) {
      let tagString = "@example";

      const captionRegex = /^\s*<caption>([\s\S]*?)<\/caption>/i;
      const exampleCaption = captionRegex.exec(description);
      if (exampleCaption) {
        description = description.slice(exampleCaption[0].length);
        tagString += ` <caption>${exampleCaption[1].trim()}</caption>`;
      }

      description = trimIndentation(trimTrailingSpaces(description));

      const EXAMPLE_CODE_INDENT = "  ";

      const examplePrintWith = printWidth - EXAMPLE_CODE_INDENT.length;

      let formattedExample: string | undefined;

      // try JSON formatting
      if (formattedExample === undefined && /^\s*\{/.test(description)) {
        formattedExample = tryFormat(description, {
          ...options,
          parser: "json",
          printWidth: examplePrintWith,
        });
      }

      // try current language formatting
      if (formattedExample === undefined) {
        formattedExample = tryFormat(description, {
          ...options,
          printWidth: examplePrintWith,
        });
      }

      // unformatted
      if (formattedExample === undefined) {
        if (jsdocKeepUnParseAbleExampleIndent) {
          formattedExample = description;
        } else {
          formattedExample = description
            .split(/\n/g)
            .map((line) => line.trim())
            .join("\n");
        }
      }

      // some formatters add empty lines
      formattedExample = trimEmptyLines(formattedExample);

      // add indentation
      formattedExample = prefixLinesWith(
        formattedExample,
        EXAMPLE_CODE_INDENT,
        true,
      );

      tagString += "\n" + formattedExample;

      return {
        value: tagString,
        insertLineAfter: true,
      };
    } else {
      let tagString = `@${tag}`;

      const padLength = (length: number): void => {
        tagString = tagString.padEnd(length, " ");
      };

      const getTypeOffset = () =>
        shouldAlign(tag) ? TYPE_OFFSET : tagString.length + jsdocSpaces;
      const getNameOffset = () =>
        shouldAlign(tag) ? NAME_OFFSET : tagString.length + jsdocSpaces;
      const getDescOffset = () =>
        shouldAlign(tag) ? DESC_OFFSET : tagString.length + jsdocSpaces;

      if (type) {
        padLength(getTypeOffset());
        tagString += `{${type}}`;
      }
      if (name) {
        padLength(getNameOffset());
        tagString += name;
      }

      if (description) {
        if (!TAGS_NEED_FORMAT_DESCRIPTION.includes(tag)) {
          padLength(getDescOffset());
          tagString += description;
        } else {
          const beforeDescWidth = shouldAlign(tag)
            ? getDescOffset()
            : getLastLine(tagString).length + jsdocSpaces;
          if (beforeDescWidth >= printWidth) {
            tagString += "\n" + formatDescription(description, options);
          } else {
            const formatted = formatDescription(description, options, {
              firstLinePrintWidth: printWidth - beforeDescWidth,
            });

            const firstLine = getFirstLine(formatted);
            if (!firstLine) {
              // empty first line
              tagString += formatted;
            } else if (beforeDescWidth + firstLine.length <= printWidth) {
              // formatted description fits within print width
              padLength(getDescOffset());
              tagString += formatted;
            } else {
              // the first line is too long
              tagString += "\n" + formatted;
            }
          }
        }
      }

      return {
        value: tagString,
        insertLineAfter: tag === TODO,
      };
    }
  }

  let result = "";
  let emptyLineAfter = false;
  tags.forEach((tag) => {
    const { value, insertLineBefore, insertLineAfter } = stringifyTag(tag);

    result += "\n";
    if (insertLineBefore && !emptyLineAfter) {
      result += "\n";
    }
    result += value;
    if (insertLineAfter) {
      result += "\n";
      emptyLineAfter = true;
    } else {
      emptyLineAfter = false;
    }
  });

  return trimEmptyLines(result);
}

export { stringifyJSDoc };
