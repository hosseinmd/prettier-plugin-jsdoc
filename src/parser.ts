import commentParser from "comment-parser";
import {
  convertToModernType,
  countLines,
  formatType,
  prefixLinesWith,
  trimTrailingSpaces,
} from "./utils";
import { DESCRIPTION } from "./tags";
import {
  TAGS_DESCRIPTION_NEEDED,
  TAGS_IS_CAMEL_CASE,
  TAGS_NAMELESS,
  TAGS_ORDER,
  TAGS_SYNONYMS,
  TAGS_TYPELESS,
} from "./roles";
import { AST, JsdocOptions, PrettierComment } from "./types";
import { stringifyJSDoc } from "./stringify";
import { convertCommentDescToDescTag } from "./descriptionFormatter";
import { Parser } from "prettier";

/** @link https://prettier.io/docs/en/api.html#custom-parser-api} */
export const getParser = (parser: Parser["parse"]) =>
  function jsdocParser(
    text: string,
    parsers: Parameters<Parser["parse"]>[1],
    options: JsdocOptions,
  ): AST {
    const ast = parser(text, parsers, options) as AST;

    if (!options.jsdocParser) {
      return ast;
    }

    /**
     * Control order of tags by weights. Smaller value brings tag higher.
     *
     * @param {String} tagTitle
     * @returns {Number} Tag weight
     */
    function getTagOrderWeight(tagTitle: string): number {
      if (tagTitle === DESCRIPTION && !options.jsdocDescriptionTag) {
        return -1;
      }
      const index = TAGS_ORDER.indexOf(tagTitle);
      return index === -1 ? TAGS_ORDER.indexOf("other") : index;
    }

    ast.comments.forEach((comment) => {
      if (!isBlockComment(comment)) return;

      /** Issue: https://github.com/hosseinmd/prettier-plugin-jsdoc/issues/18 */
      comment.value = comment.value.replace(/^([*]+)/g, "*");

      const commentString = `/*${comment.value}*/`;

      /**
       * Check if this comment block is a JSDoc. Based on:
       * https://github.com/jsdoc/jsdoc/blob/master/packages/jsdoc/plugins/commentsOnly.js
       */
      if (!/^\/\*\*[\s\S]+?\*\/$/.test(commentString)) return;

      const parsed = commentParser(commentString, {
        dotted_names: false,
        trim: false,
      })[0];

      if (!parsed) {
        // Error on commentParser
        return;
      }

      const commentIndentationWidth = getIndentationWidth(comment, options);
      const CONTENT_PREFIX = " * ";
      const commentContentPrintWidth =
        options.printWidth - commentIndentationWidth - CONTENT_PREFIX.length;

      convertCommentDescToDescTag(parsed);

      const tags = parsed.tags
        // Prepare tags data
        .map(
          ({
            name,
            description,
            type,
            tag,
            source,
            optional,
            default: _default,
            ...restInfo
          }) => {
            name = (name || "").trim();
            description = trimTrailingSpaces((description || "").trim());

            /** When space between tag and type missed */
            const tagSticksToType = tag.indexOf("{");
            if (tagSticksToType !== -1 && tag[tag.length - 1] === "}") {
              type = tag.slice(tagSticksToType + 1, -1);
              tag = tag.slice(0, tagSticksToType);
            }

            if (TAGS_ORDER.includes(tag) && !TAGS_IS_CAMEL_CASE.includes(tag)) {
              tag = tag && tag.trim().toLowerCase();
            }
            if (tag in TAGS_SYNONYMS) {
              tag = TAGS_SYNONYMS[tag as keyof typeof TAGS_SYNONYMS];
            }

            if (TAGS_NAMELESS.includes(tag) && name) {
              description = `${name} ${description}`;
              name = "";
            }
            if (TAGS_TYPELESS.includes(tag) && type) {
              description = `{${type}} ${description}`;
              type = "";
            }

            if (type) {
              /**
               * Convert optional to standard
               * https://jsdoc.app/tags-type.html#:~:text=Optional%20parameter
               */
              type = type.replace(/[=]$/, () => {
                optional = true;
                return "";
              });

              type = convertToModernType(type);
              type = formatType(type, {
                ...options,
                printWidth: commentContentPrintWidth,
              });

              // Additional operations on name
              if (name) {
                // Optional tag name
                if (optional) {
                  // Figure out if tag type have default value
                  _default = _default?.trim();
                  if (_default) {
                    description = description
                      .trim()
                      .replace(/[ \t]*Default is `.*`\.?$/, "");
                    if (description && !/[.\r\n]$/.test(description)) {
                      description += ".";
                    }
                    description += ` Default is \`${_default}\``;
                    name = `[${name}=${_default}]`;
                  } else {
                    name = `[${name}]`;
                  }
                }
              }
            }

            description = description.trim();

            return {
              ...restInfo,
              name,
              description,
              type,
              tag,
              source,
              default: _default,
              optional,
            } as commentParser.Tag;
          },
        )
        // Sort tags
        .sort((a, b) => getTagOrderWeight(a.tag) - getTagOrderWeight(b.tag))
        .filter(({ description, tag }) => {
          if (!description && TAGS_DESCRIPTION_NEEDED.includes(tag)) {
            return false;
          }
          return true;
        });

      const commentContent = stringifyJSDoc(tags, {
        ...options,
        printWidth: commentContentPrintWidth,
      });

      if (commentContent) {
        if (
          countLines(commentContent) === 1 &&
          commentIndentationWidth + `/**  */`.length + commentContent.length <=
            options.printWidth
        ) {
          comment.value = `* ${commentContent} `;
        } else {
          comment.value = `*\n${prefixLinesWith(
            commentContent,
            CONTENT_PREFIX,
          )}\n `;
        }
      } else {
        comment.value = "";
      }
    });

    ast.comments = ast.comments.filter(
      (comment) => !isBlockComment(comment) || comment.value,
    );

    return ast;
  };

function isBlockComment(comment: PrettierComment): boolean {
  return comment.type === "CommentBlock" || comment.type === "Block";
}

function getIndentationWidth(
  comment: PrettierComment,
  options: JsdocOptions,
): number {
  let width = comment.loc.start.column;
  if (options.useTabs) {
    width *= options.tabWidth;
  }
  return width;
}
