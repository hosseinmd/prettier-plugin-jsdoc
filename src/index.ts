import { getParser } from "./parser";
import parserBabel from "prettier/parser-babel";
import parserFlow from "prettier/parser-flow";
import parserTypescript from "prettier/parser-typescript";
import prettier, { SupportOption } from "prettier";
import { JsdocOptions } from "./types";
import { findPluginByParser } from "./utils";

const options: Record<keyof JsdocOptions, SupportOption> = {
  jsdocSpaces: {
    name: "jsdocSpaces",
    type: "int",
    category: "jsdoc",
    default: 1,
    description: "How many spaces will be used to separate tag elements.",
  },
  jsdocDescriptionWithDot: {
    name: "jsdocDescriptionWithDot",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should dot be inserted at the end of description",
  },
  jsdocDescriptionTag: {
    name: "jsdocDescriptionTag",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should description tag be used",
  },
  jsdocVerticalAlignment: {
    name: "jsdocVerticalAlignment",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should tags, types, names and description be aligned",
  },
  jsdocKeepUnParseAbleExampleIndent: {
    name: "jsdocKeepUnParseAbleExampleIndent",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description:
      "Should unParseAble example (pseudo code or no js code) keep its indentation",
  },
  jsdocSingleLineComment: {
    name: "jsdocSingleLineComment",
    type: "boolean",
    category: "jsdoc",
    default: true,
    description: "Should compact single line comment",
  },
  jsdocSeparateReturnsFromParam: {
    name: "jsdocSeparateReturnsFromParam",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Add an space between last @param and @returns",
  },
  jsdocCapitalizeDescription: {
    name: "jsdocCapitalizeDescription",
    type: "boolean",
    category: "jsdoc",
    default: true,
    description: "Should capitalize first letter of description",
  },
  tsdoc: {
    name: "tsdoc",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should format as tsdoc",
  },
  jsdocPrintWidth: {
    name: "jsdocPrintWidth",
    type: "int",
    category: "jsdoc",
    default: undefined as any,
    description:
      "If You don't set value to jsdocPrintWidth, the printWidth will be use as jsdocPrintWidth.",
  },
};

const defaultOptions: JsdocOptions = {
  jsdocSpaces: options.jsdocSpaces.default as number,
  jsdocPrintWidth: (options.jsdocPrintWidth.default as unknown) as undefined,
  jsdocDescriptionWithDot: options.jsdocDescriptionWithDot.default as boolean,
  jsdocDescriptionTag: options.jsdocDescriptionTag.default as boolean,
  jsdocVerticalAlignment: options.jsdocVerticalAlignment.default as boolean,
  jsdocKeepUnParseAbleExampleIndent: options.jsdocKeepUnParseAbleExampleIndent
    .default as boolean,
  jsdocSingleLineComment: options.jsdocSingleLineComment.default as boolean,
  jsdocSeparateReturnsFromParam: options.jsdocSeparateReturnsFromParam
    .default as boolean,
  jsdocCapitalizeDescription: options.jsdocCapitalizeDescription
    .default as boolean,
  tsdoc: options.tsdoc.default as boolean,
};

const languages = prettier
  .getSupportInfo()
  .languages.filter(({ name }) =>
    [
      "JavaScript",
      "Flow",
      "JSX",
      "TSX",
      "TypeScript",
      "Markdown",
      "MDX",
    ].includes(name),
  );

const parsers = {
  // JS - Babel
  get babel() {
    const parser = parserBabel.parsers.babel;
    return mergeParsers(parser, "babel");
  },
  get "babel-flow"() {
    const parser = parserBabel.parsers["babel-flow"];
    return mergeParsers(parser, "babel-flow");
  },
  get "babel-ts"() {
    const parser = parserBabel.parsers["babel-ts"];
    return mergeParsers(parser, "babel-ts");
  },
  // JS - Flow
  get flow() {
    const parser = parserFlow.parsers.flow;
    return mergeParsers(parser, "flow");
  },
  // JS - TypeScript
  get typescript(): prettier.Parser {
    const parser = parserTypescript.parsers.typescript;

    return mergeParsers(parser, "typescript");
    // require("./parser-typescript").parsers.typescript;
  },
  get "jsdoc-parser"() {
    // Backward compatible, don't use this in new version since 1.0.0
    const parser = parserBabel.parsers["babel-ts"];

    return mergeParsers(parser, "babel-ts");
  },
};

function mergeParsers(originalParser: prettier.Parser, parserName: string) {
  const jsDocParse = getParser(originalParser.parse, parserName) as any;

  const jsDocPreprocess = (text: string, options: prettier.ParserOptions) => {
    const tsPluginParser = findPluginByParser(parserName, options);

    if (!tsPluginParser) {
      return originalParser.preprocess
        ? // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          //@ts-ignore
          originalParser.preprocess(text, options)
        : text;
    }

    const preprocess = tsPluginParser.preprocess || originalParser.preprocess;

    Object.assign(parser, {
      ...parser,
      ...tsPluginParser,
      preprocess: jsDocPreprocess,
      parse: jsDocParse,
    });

    return preprocess ? preprocess(text, options) : text;
  };

  const parser = {
    ...originalParser,
    preprocess: jsDocPreprocess,
    parse: jsDocParse,
  };

  return parser;
}

export { languages, options, parsers, defaultOptions };
