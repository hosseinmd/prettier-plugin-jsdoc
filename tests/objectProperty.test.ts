import * as prettier from "prettier";
import { AllOptions } from "../src/types";

function subject(code: string, options: Partial<AllOptions> = {}) {
  return prettier.format(code, {
    parser: "babel-flow",
    plugins: ["prettier-plugin-jsdoc"],
    jsdocSpaces: 1,
    ...options,
  } as AllOptions);
}

test("object property", async () => {
  const result = await subject(`

/**	
 * Copyright (c) 2015-present, Facebook, Inc.	
 * All rights reserved.	
 *   
 *    This source code is licensed under the license found in the LICENSE file in	
 * the root directory of this source tree.	
 */	

/**	
 * Copyright (c) 2015-present, Facebook, Inc.	
 * All rights reserved.	
 *   
 * This source code is licensed under the license found in the LICENSE file in	
 * the root directory of this source tree.	
 */	

const obj = {
  /**
* @param {object} [filters]
   * @param {string} [filters.searchInput]
   * @param {boolean} [filters.isActive]
   * @param {boolean} [filters.isPerson]
   * @param {import('../types').IdentityStatus} [filters.identityStatuses]
   * @param {string} [filters.lastActivityFrom] YYYY-MM-DD
   * @param {string} [filters.lastActivityTo]
   * @param {string} [filters.registeredFrom]
   * @param {string} [filters.registeredTo]
   * @param {number} [filters.skip]
   * @param {number} [filters.take]
   * @param {string} [filters.orderBy]
   * @param {boolean} [filters.orderDesc]
   * @returns {import('axios').AxiosResponse<
    import('../types').ResellerUserIntroduced[]
  >}
  */
  a(filters) {},
};
`);

  expect(result).toMatchSnapshot();
});
