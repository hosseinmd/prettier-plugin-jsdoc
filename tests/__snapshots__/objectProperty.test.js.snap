// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`object property 1`] = `
"/**
 * @providesModule Foo
 * @providesModule bar
 * @flow
 */

/**
 * Copyright (c) 2015-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

/**
 * Copyright (c) 2015-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

const obj = {
  /**
   * @param {object} [filters]
   * @param {string} [filters.searchInput]
   * @param {boolean} [filters.isActive]
   * @param {boolean} [filters.isPerson]
   * @param {import(\\"../types\\").IdentityStatus} [filters.identityStatuses]
   * @param {string} [filters.lastActivityFrom] YYYY-MM-DD
   * @param {string} [filters.lastActivityTo]
   * @param {string} [filters.registeredFrom]
   * @param {string} [filters.registeredTo]
   * @param {number} [filters.skip]
   * @param {number} [filters.take]
   * @param {string} [filters.orderBy]
   * @param {boolean} [filters.orderDesc]
   * @returns {import(\\"axios\\").AxiosResponse<
   *   import(\\"../types\\").ResellerUserIntroduced[]
   * >}
   */
  a(filters) {},
};
"
`;
